rcl_action/test/rcl_action/test_goal_handle.cpp

```diff
@@ -0,0 +1,300 @@
+// limitations under the License.
+#include <gtest/gtest.h>
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rcl_action/goal_handle.h"
+#include "rcl_action/types.h"
+
+#include "rcl/error_handling.h"
+
+TEST(TestGoalHandle, test_goal_handle_init_fini)
+{
+  rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+
+  // Initialize with a null goal handle
+  rcl_ret_t ret = rcl_action_goal_handle_init(nullptr, &goal_info, rcl_get_default_allocator());
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with a null goal info
+  rcl_action_goal_handle_t goal_handle = rcl_action_get_zero_initialized_goal_handle();
+  EXPECT_EQ(goal_handle.impl, nullptr);
+  ret = rcl_action_goal_handle_init(&goal_handle, nullptr, rcl_get_default_allocator());
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with an invalid allocator
+  rcl_allocator_t invalid_allocator = (rcl_allocator_t)rcutils_get_zero_initialized_allocator();
+  ret = rcl_action_goal_handle_init(&goal_handle, &goal_info, invalid_allocator);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Initialize with valid goal handle and allocator
+  ret = rcl_action_goal_handle_init(&goal_handle, &goal_info, rcl_get_default_allocator());
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  EXPECT_NE(goal_handle.impl, nullptr);
+
+  // Try to initialize again
+  ret = rcl_action_goal_handle_init(&goal_handle, &goal_info, rcl_get_default_allocator());
+  EXPECT_EQ(ret, RCL_RET_ALREADY_INIT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Finalize with null goal handle
+  ret = rcl_action_goal_handle_fini(nullptr);
+  EXPECT_EQ(ret, RCL_RET_INVALID_ARGUMENT) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Finalize with valid goal handle
+  ret = rcl_action_goal_handle_fini(&goal_handle);
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+}
+
+TEST(TestGoalHandle, test_goal_handle_is_valid)
+{
+  // Check null goal handle
+  bool is_valid = rcl_action_goal_handle_is_valid(nullptr);
+  EXPECT_FALSE(is_valid) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Check uninitialized goal handle
+  rcl_action_goal_handle_t goal_handle = rcl_action_get_zero_initialized_goal_handle();
+  is_valid = rcl_action_goal_handle_is_valid(&goal_handle);
+  EXPECT_FALSE(is_valid) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  // Check valid goal handle
+  rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+  rcl_ret_t ret = rcl_action_goal_handle_init(
+    &goal_handle, &goal_info, rcl_get_default_allocator());
+  EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  is_valid = rcl_action_goal_handle_is_valid(&goal_handle);
+  EXPECT_TRUE(is_valid) << rcl_get_error_string().str;
+}
```

Comment:
Recommend `_fini()` on the goal handle to remove a false positive when running tests with address sanitizer.

Reply:
Added d1fa431